src/drive/lib/encryption.js

```diff
@@ -1,3 +1,80 @@
+const stringToArrayBuffer = string => {
+  var encoder = new TextEncoder('utf-8')
+  return encoder.encode(string)
+}
+
+const asDerivableKey = async (
+  baseMaterial,
+  keyDerivationAlgorithm = 'PBKDF2'
+) => {
+  let usableMaterial
+  if (baseMaterial instanceof CryptoKey) {
+    usableMaterial = await window.crypto.subtle.exportKey('raw', baseMaterial)
+  } else {
+    usableMaterial = stringToArrayBuffer(baseMaterial)
+  }
+  return await window.crypto.subtle.importKey(
+    'raw',
+    usableMaterial,
+    { name: keyDerivationAlgorithm },
+    true,
+    ['deriveKey']
+  )
+}
+
+const slowHashing = async (
+  baseKey,
+  saltBuffer,
+  { keyDerivationAlgorithm, iterations, hash } = {},
+  { algorithm, keyLength } = {}
+) => {
+  return await window.crypto.subtle.deriveKey(
+    {
+      name: keyDerivationAlgorithm || 'PBKDF2',
+      salt: saltBuffer,
+      iterations: iterations || 1000,
+      hash: hash || 'SHA-256'
+    },
+    baseKey,
+    {
+      name: algorithm || 'AES-GCM',
+      length: keyLength || 256
+    },
+    true,
+    ['encrypt', 'decrypt', 'wrapKey', 'unwrapKey']
+  )
+}
+
+export const deriveKey = async (password, salt) => {
+  const passwordAsKey = await asDerivableKey(password)
+  const saltBuffer = stringToArrayBuffer(salt)
+  const passwordBuffer = stringToArrayBuffer(password)
+  // Chain 2 key derivations : first, derive a key from a password
+  const preKey = await slowHashing(passwordAsKey, saltBuffer, {
+    iterations: 1000
+  })
+  const key = await asDerivableKey(preKey).then(key => {
+    return slowHashing(key, passwordBuffer, { iterations: 1 })
```

Comment:
The more iterations, the more secure? 1 might be risky?

Reply:
This is the 2nd derivation step, occurring after the first 100k rounds of PBKDF2. This is how we planed to proceed in our Paper according to 1Password's scheme.

Reply:
Right, my bad